### ISSUE TITLE
Improve Nifty-Gui documentation

### ISSUE THREAD

**Issue by relu91** [nifty-gui/nifty-gui#272]
Hello, the title is a bit ambiguous, nifty-gui actually has a great documentation but it suffers to be a lot static in the time. This is caused by that the only source of satisfying knowledge about nifty is locked in a pdf. As I mentioned in #271 at least an editable version (.doc) in the git repository would helps. People could make pr with their changes. An other choice would be to import the whole manual to github wiki but it's a time consuming work since I think we should import page by page. There's an other option, in my opinion, that is to let the manual an github wiki to coexist, giving different scopes to each. Maybe the wiki will cointain only high level things or noob guides wile the manual will have more deep content. Or even the wiki could treat about nifty-gui development process , with pages like "How rendering works". I know that there are more urgent things to do but you can see this issue as a reference for the future :+1: . I hope to have a nice discussion here.

**Comment by oberien** [comment]
I would prefer a well organized wiki, but as you said, that would be a lot of work. Maybe someone of the community could import it into a wiki, maybe even an extern wiki and not the github-wiki (e.g. wikia).
I would only agree to a coexistance, if the manual was used as an introduction into nifty as kind of a tutorial. But in a wiki I'd like to find everything :) From the basics (maybe someone can integrate an tutorial into the wiki) over standard methods to use (e.g. issue 271 - getBackgroundColor and setBackgroundColor) to maybe even in-depth descriptions about the rendering and font-loading etc.

**Comment by ghost** [comment]
Also, the JME wiki could be used (at least in the beginning) for basic tutorials (at least at the moment), as the basics of Nifty are all pretty much covered up in it.

**Comment by ghost** [comment]
Manuals are dead. Everything **_needs**_ to be moved over to the wiki. That's the only long term solution that makes sense. Nifty Gui is still undergoing so much change, that the only way to keep up is to use a wiki because of it's dynamic nature. Even if you have an editable manual document, when Nifty Gui is released, the manual everyone gets with the release version will be frozen.

So, for example. say that there's not clear documentation about a specific feature in the manual when the next version of Nifty Gui is released. Then someone wants to add that information. If they add it to a manual, then the "release" users won't know about it until the next release version of Nifty Gui. They will come and create an issue on GitHub, when we'll say, "well, it's in the manual", they'll say, "well it's not in my version".

This kind of pain is what wiki's were created to solve.

**Comment by ghost** [comment]
Not to mention that the manual exists in the source code, it's not easy to find if you don't build Nifty Gui from source. Everyone nowadays expects the information to be in a wiki. No one reads manuals any more lol.

**Comment by relu91** [comment]
Yeah that's right... I thought that the manual could represent the same thing of JME beginner's guide but as you said nifty-gui is far more dynamic than jme so it could be a pain. I was thinking how can we handle the difference in nifty-gui version on the wiki ? Different section ? or specific paragraphs in a general one ? ( like in the manual when it's talking about margins ) 

> Also, the JME wiki could be used (at least in the beginning) for basic tutorials (at least at the moment), as the basics of Nifty are all pretty much covered up in it.

yeah there are a lot of useful tips there but are most concerned about nifty-gui jme integration. we should provide documentation to all supported renderer or environment.

**Comment by ghost** [comment]
> relu91 commented 14 minutes ago
> 
> I was thinking how can we handle the difference in nifty-gui version on the wiki ? Different section ? or specific paragraphs in a general one ? ( like in the manual when it's talking about margins )

The main content of the wiki should be assumed to be the latest released version, i.e., 1.4. Then there should be separate sections for "odd" versions, such as 1.4-SNAPSHOT (experimental), 1.3 (outdated / deprecated), 2.0 (dreamland), etc. Those "odd" sections should not repeat all the main content, only describe features that are specific to / different from the main current release.

It's a crime against humanity to have notes scattered throughout the manual about different versions (ugly and confusing, nearly killing readability).

The main wiki should specify on its home page that it's for 1.4 release version, with links to differences for other versions.

**Comment by relu91** [comment]
> It's a crime against humanity to have notes scattered throughout the manual about different versions (ugly and confusing, nearly killing readability). 

:laughing: . About the main content we should follow the some structure of the manual or do you think we can have more freedom ?

**Comment by ghost** [comment]
Version 2.0 will probably require it's own wiki because it's so different.

**Comment by ghost** [comment]
> relu91 commented 19 seconds ago
> 
> :laughing: . About the main content we should follow the some structure of the manual or do you think we can have more freedom ?

I don't really care for the layout in the manual, I think it's pretty disorganized and we could do way better. There's no point in copying the structure. During transcription we should aim to re-organize it. A wiki is not a PDF so it shouldn't be organized the same way (but the manual's organization kind of sucks anyway). The manual should simply be viewed as a giant box of content that we can pull from, rather than a 1:1 copy.

**Comment by ghost** [comment]
I'm willing to work on the transcription, not only am I excited about it, but I have a lot of content to add myself that is not even in the manual or wiki yet.

**Comment by ghost** [comment]
I need to add a chapter on LibGDX integration, batch rendering configuration, deprecated rendering methods, how to add a new backend to Nifty Gui, and a section on the future of Nifty Gui, that can be a space for saving ideas, like preparing Nifty Gui for Android, moving the rendering system forward, integration with https://github.com/void256/coregl, and missing / needed / undocumented features, such as a scrollpanel-based chat box, non-continuous rendering, multithreading issues, effects where no documentation exists, and on and on and on....

**Comment by ghost** [comment]
Thinking about starting on it right now actually... I'm bored of hacking on my [game](http://forums.tigsource.com/index.php?topic=42666.0).

**Comment by relu91** [comment]
I'm eager to read all that things! :+1: . I'm glad that you are starting to work with this I think that it would boost the nifty-gui development! . Just a little of topic thing ... the libgdx integrant can't run on android ? sorry for dumb question but the new batch system is totally obscure to me :) I thought that you are using the ligbdx render engine to render nifty-gui.

**Comment by ghost** [comment]
> relu91 commented a minute ago
> 
> I'm eager to read all that things! :+1: . I'm glad that you are starting to work with this I think that it would boost the nifty-gui development! . Just a little of topic thing ... the libgdx integrant can't run on android ? sorry for dumb question but the new batch system is totally obscure to me :) I thought that you are using the ligbdx render engine to render nifty-gui.

Yes, Nifty Gui can run on Android via nifty-renderer-libgdx. It just needs some tweaking, and a nice nifty-examples-mobile module with tiny assets to demonstrate it. I got one of the nifty-examples demos running on my old Android phone like a year ago, but I haven't had time to finish that project.

**Comment by ghost** [comment]
Well maybe I should post here when I update the wiki? This could be a never ending issue ;-).

https://github.com/void256/nifty-gui/wiki/Working-from-Source (near complete rewrite)
https://github.com/void256/nifty-gui/wiki/Examples (new)
https://github.com/void256/nifty-gui/wiki/Controls (updated wording & links, converted completely to markdown)

I think all the wiki pages should be the same format. Right now some are mediawiki, some are markdown. I prefer markdown and am most familiar with it, so if I'm gonna be writing a lot of wiki, well they better be markdown ;-).

**Comment by relu91** [comment]
Yeah it's better with only markdown. The reason why it was built with media wiki it's that I had simply cut&paste the pages from the old wiki without to rewrite them from scratch! So awesome work! About posting here, you could refer this issue when you edit the wiki I think it works.

**Comment by Benjamin-Justice** [comment]
Hi, does the wiki allow for branches?
Because I love the "feature" that the manual I have is frozen for Nifty 1.3.2.

If you have the manual in tex format, or any "versionable" format I can edit without destroying the layout (which means no .doc), I will gladly assist you for Nifty 1.3.2 as it is the version integrated in jMonkey 3.0.
It would help to myself and my team.

At the moment I use the source code, manual and sometimes wiki for my doc needs, which is nasty tbh.
Also links in the manual when another section is referred would be magic.
